### ISSUE TITLE
Match service results explanation

### ISSUE THREAD

**Issue by vitalidze** [Project-OSRM/osrm-backend#2949]
Hi

First of all, thank you for this project. I am a developer of another open-source project, which was using the v4 API and now I think it's time to migrate to v5. However, the response format is changed and I don't understand how to achieve same results as I did with v4.

Usually I was doing a request with a set of latitude/longitude and their timestamps. v4 returned me a set of matchings with indices from original array, matched points and suggested geometry. So then I can associate matched points with original points, combine track segments geometries (obtained from matchings) and draw a line on map.

Now, for example for this query:

https://router.project-osrm.org/match/v1/driving/37.297108,55.452800;37.297340,55.452968;37.297440,55.453348;37.297432,55.453456;37.294388,55.455488;37.294412,55.455664;37.295116,55.456724;37.296848,55.459384;37.298572,55.462020;37.300276,55.464664;37.301864,55.466788;37.304888,55.468972;37.308008,55.471048;37.309760,55.472216;37.312368,55.473956;37.315464,55.476040;37.317792,55.477684;37.319540,55.480216;37.321312,55.482912;37.322556,55.484800;37.324236,55.487348;37.325932,55.489924;37.327000,55.491544;37.328680,55.494132;37.330428,55.496776;37.332108,55.499376;37.333716,55.501536;37.335024,55.502548;37.338788,55.504388;37.342704,55.506112;37.346288,55.507692;37.350160,55.509392;37.354060,55.511100;37.358040,55.512852;37.361940,55.514560;37.365928,55.516252;37.370188,55.517612;37.374768,55.518900;37.379172,55.520148;37.380660,55.520716;37.382528,55.521744;37.385276,55.524048;37.387980,55.526496;37.389112,55.528204;37.389400,55.529168;37.389888,55.531940;37.391196,55.534288;37.393060,55.536356;37.394320,55.537216;37.398064,55.539040;37.402112,55.540672;37.406120,55.542284;37.410020,55.543864;37.411004,55.544260;37.414628,55.545736;37.418668,55.547372;37.422648,55.548984;37.425116,55.549984;37.425844,55.550276;37.427568,55.550976;37.428828,55.551600;37.429348,55.551904?timestamps=1414129581;1414129905;1414129941;1414129946;1414129983;1414129987;1414129998;1414130016;1414130032;1414130048;1414130061;1414130077;1414130092;1414130122;1414130169;1414130198;1414130219;1414130235;1414130250;1414130262;1414130283;1414130302;1414130332;1414130382;1414130402;1414130421;1414130439;1414130448;1414130466;1414130485;1414130515;1414130542;1414130562;1414130581;1414130597;1414130614;1414130630;1414130644;1414130658;1414130663;1414130670;1414130683;1414130698;1414130707;1414130712;1414130726;1414130737;1414130749;1414130755;1414130769;1414130785;1414130803;1414130828;1414130858;1414130895;1414130913;1414130930;1414130961;1414130991;1414131021;1414131052;1414131082

There are 49 points in the query, but it returns me two matchings and a weird array of waypoints. The "weird" array contains holes (`null`) representing something. Then result contains about 61 waypoints. I don't understand how I can create an relation between set of original points and the set from reply. Can you please explain what all these mean?

**Comment by daniel-j-h** [comment]
Here's a great explanation from a couple of days ago: https://github.com/Project-OSRM/osrm-backend/issues/2933#issuecomment-249068441

**Comment by vitalidze** [comment]
@daniel-j-h thank you for answering. However, it does not help. Let's get back to my request and response to the request. In reply I am getting two matchings: one with one leg, and another one with multiple legs. Then when I walk through `tracepoints` I am seeing the following:

````json
 "tracepoints": [
    null,
    null,
    null,
    {
      "waypoint_index": 0,
      "location": [
        37.297571,
        55.453531
      ],
      "name": "",
      "hint": "RYGLA0iBi4MAAAAAAwEAAF4AAAAAAAAAtAEAAKgt3gJhLt4CCrUAAKMdOQJbJ04DGB05AhAnTgMAAAEBqK3C_A==",
      "matchings_index": 0
    },
    {
      "waypoint_index": 1,
      "location": [
        37.29429,
        55.45543
      ],
      "name": "",
      "hint": "QYGLA0SBi4MAAAAAGQAAAFMAAAAAAAAAAAAAADredAo43nQKCrUAANIQOQLGLk4DNBE5AgAvTgMAAAEBqK3C_A==",
      "matchings_index": 0
    },
    null,
    null,
    null,
    null,
    null,
    {
    "waypoint_index": 0,
      "location": [
        37.301671,
        55.466862
      ],
      "name": "Калужское шоссе",
      "hint": "YtXOA29l94X0zq0AIgAAAAoAAAAAAAAAqgAAAOLSLgOH1N4ACrUAAKctOQJuW04DaC45AiRbTgMAAAEBqK3C_A==",
      "matchings_index": 1
    },
````

... then goes all positions for `matching` 2. I am confused with:

* many `null`s for the first matching (or for the second too?)
* the matchings **do** intersect

What should I do with such response? How to draw a track matched to the road maps?

**Comment by TheMarex** [comment]
@vitalidze 
- `result.tracepoints` is an array that translates between input coordinates and matched coordinates. A `null` entry means the input coordinate could not be matched.
- `result.tracepoints[i].matchings_index` is the index to `result.matchings` that the input coordinate corresponds to.
- `result.tracepoints[i].waypoint_index` is the index of input trace point in the resulting sub-matching
- `result.matchings` is an array of sub-matchings. Each sub-matching corresponds of multiple legs (path between matched trace points). If you just want to draw a geometry use `result.matchings[i].geometrie` and put that on a map.

**Comment by vitalidze** [comment]
@TheMarex well, thank you for spending time on the explanation. But it does not add any new information. 

My task is not just to draw geometries, but also to replace original points with matched points so I can draw original points on the newly received geometry. And here I become confused because for example for point 0 I have two matched points.

**Comment by TheMarex** [comment]
>  And here I become confused because for example for point 0 I have two matched points.

Can you explain this a little more? I don't get what you mean by "two matched points". `tracepoints[0]` is `null` meaning it was not matched.

**Comment by vitalidze** [comment]
I am sending a query for 49 points, reply contains about 61 points. Then I have `tracepoints[3]` with `waypoint_index` = 0, `location` = 37.297571, 55.453531 and also `trancepoints[10]` with same `waypoint_index` = 0, `location` = 37.301671, 55.466862. According to the documentation and your reply, both these points are for the point with index = 0 from the input query. So, my question - which one I should take?

**Comment by TheMarex** [comment]
@vitalidze ah sorry I think you misunderstood me. :-) `waypoint_index`is the index to the leg in `result.matchings[matchings_index].legs[waypoint_index]` where the matched point is the start point. If you have a look at `matchings_index` you will see that it is different for both points.

**Comment by vitalidze** [comment]
Sure I am seeing that the `matchings_index` is different. So how can I find relation between input points (from the HTTP query) and matched points? I am asking because I am trying to migration from v4 to v5 of OSRM API, where it was obvious.

**Comment by TheMarex** [comment]
@vitalidze well it is even easier? Say you have input coordinates from 0 to 10. Then `tracepoints[0].location` to `tracepoints[10].location` is your matched coordinates? `tracepoints` is index the same way as you input coordinate array. Just like the `waypoints` array in other API responses.

**Comment by vitalidze** [comment]
Well, in my case, input array length was different from the `tracepoints` length. This is why I got confused. However, I have checked right now and it's same for this particular query. Sorry about that. 

The thing is that I have already written code, that expects matching as you describe, but it was failing due to a difference in input and output arrays lengths. Then I have created this thread.

I will get back once I have any update on this. Thank you for your time!
